### PR TITLE
Prevent Consuming Unconfirmed UTXOs if tx has a memo

### PR DESCRIPTION
**Pull request by thorswapdev**
Issue: https://github.com/xchainjs/xchainjs-lib/issues/330

# CHANGELOG

## Package

`@xchainjs/xchain-bitcoin`

### Breaking change

- prevent spending unconfirmed UTXOs
- update `client.transfer()` to pass `spendPendingUTXO` param to the `Utils.buildTx()` 
- update `Utils.buildTx()` to spend only confirmed UTXO if memo exists
- update `Utils.buildTx()` to build `psbt` using `accumulative` lib
- add `getIsTxConfirmed`, `getConfirmedUnspentTxs` sochain api
- add unit test for sochain apis
- add unit test for both success and failed cases of `client.transfer()` to prevent spending unconfirmed UTXOs
- add `accumulative` to devDependency and peerDependency

### CODE REVIEW

**Comment by mikewyszinski** [comment]
@thorswapdev hey it looked like you pulled from chnage over from another pull request i worked on. you should probably also include the refactored [getunspenttxs()](https://github.com/xchainjs/xchainjs-lib/blob/c8dd787760bdba7022df1a01353e3ddea32e43ed/packages/xchain-bitcoin/src/sochain-api.ts#L98) so you make sure you fetch ALL UTXOs

**Comment by mikewyszinski** [comment]
@thorswapdev @veado @thorchain-admin  you know you might just want to consider merging https://github.com/xchainjs/xchainjs-lib/pull/324 into this PR....it's really a very similar issue and 324 also fixes BCH & LTC as well

**Comment by thorswapdev** [comment]
> @thorswapdev @veado @thorchain-admin you know you might just want to consider merging #324 into this PR....it's really a very similar issue and 324 also fixes BCH & LTC as well

@mikewyszinski , 
Yes, it makes more sense to merge both PR since it has a similar change for `xchainjs-bitcoin`. 
Let me know if #324 is ready for merge, and I will get that PR merged into this branch.

**Comment by mikewyszinski** [comment]
@thorswapdev  #324 is ready to merge, it fixes BTC,BCH,LTC. I suspect your fix for BTC might also be applicable for LTC, BCH since they too have utxos. let me know if you want me to assist or if you have any concerns :-)

**Comment by thorswapdev** [comment]
> s

@mikewyszinski , I merged #324 PR and fixed conflicts. 
1. Unit tests are covered for all updates
2. QA tested with Thorswap and confirmed:
- Prevent Spending Unconfirmed UTXO if tx has a memo
- Allow spending Unconfirmed UTXO if tx doesn't have a memo
- Prevent consuming dust attack for tx with memo

**Comment by thorswapdev** [comment]
> @thorswapdev #324 is ready to merge, it fixes BTC,BCH,LTC. I suspect your fix for BTC might also be applicable for LTC, BCH since they too have utxos. let me know if you want me to assist or if you have any concerns :-)

@mikewyszinski , The same update would be needed for LTC and BCH chains and you can work on that in a separate PR after merging this update for bitcoin package only.

File: packages/xchain-litecoin/src/utils.ts [hunk 1]
```diff
@@ -112,26 +115,26 @@ export const getBalance = async (params: AddressParams): Promise<Balance[]> => {
...
- *
- * @param {GetChangeParams} params
- * @returns {number} The change amount.
- */
-const getChange = async ({ valueOut, sochainUrl, network, address }: GetChangeParams): Promise<number> => {
-  try {
-    const balances = await getBalance({ sochainUrl, network, address })
-    const [ltcBalance] = balances.filter((balance) => assetToString(balance.asset) === assetToString(AssetLTC))
-    let change = 0
+// /**
```
Comment (davidp94) [inline packages/xchain-litecoin/src/utils.ts:118]: To be removed?

**Review by davidp94 — APPROVED** [review]
Looks good to me overall;

File: packages/xchain-bitcoin/src/utils.ts [hunk 2]
```diff
@@ -2,18 +2,26 @@ import * as Bitcoin from 'bitcoinjs-lib' // https://github.com/bitcoinjs/bitcoin
 import * as sochain from './sochain-api'
 import * as blockStream from './blockstream-api'
 import { Address, Balance, Fees, Network, TxHash, TxParams } from '@xchainjs/xchain-client'
-import { assetAmount, AssetBTC, assetToBase, assetToString, BaseAmount, baseAmount } from '@xchainjs/xchain-util'
-import { AddressParams, BtcAddressUTXOs } from './types/sochain-api-types'
-import { FeeRate, FeeRates, FeesWithRates, GetChangeParams } from './types/client-types'
+import { assetAmount, AssetBTC, assetToBase, BaseAmount, baseAmount } from '@xchainjs/xchain-util'
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
```
Comment (mikewyszinski) [inline packages/xchain-bitcoin/src/utils.ts:8]: duplicated import. will push fix

File: packages/xchain-litecoin/src/utils.ts (hunk 1 above)
Comment (mikewyszinski) [reply packages/xchain-litecoin/src/utils.ts:118]: yes, may as well since no one is currently calling it

**Review by mikewyszinski — APPROVED** [review]
I cleaned up up a little and  bumped the versions and package for BCH & LTC as well.

**Pull request merged**